openml/flows/flow.py

```diff
@@ -1,80 +1,352 @@
-        self.external_version = 'sklearn_' + sklearn.__version__
 
-    def _generate_flow_xml(self):
+        for variable, variable_name in [
+                [components, 'components'],
+                [parameters, 'parameters'],
+                [parameters_meta_info, 'parameters_meta_info']]:
+            if not isinstance(variable, OrderedDict):
+                raise TypeError('%s must be of type OrderedDict, '
+                                'but is %s.' % (variable_name, type(variable)))
+
+        self.components = components
+        self.parameters = parameters
+        self.parameters_meta_info = parameters_meta_info
+        self.class_name = class_name
+
+        keys_parameters = set(parameters.keys())
+        keys_parameters_meta_info = set(parameters_meta_info.keys())
+        if len(keys_parameters.difference(keys_parameters_meta_info)) > 0:
+            raise ValueError('Parameter %s only in parameters, but not in'
+                             'parameters_meta_info.' %
+                             str(keys_parameters.difference(
+                                 keys_parameters_meta_info)))
+        if len(keys_parameters_meta_info.difference(keys_parameters)) > 0:
+            raise ValueError('Parameter %s only in parameters_meta_info, '
+                             'but not in parameters.' %
+                             str(keys_parameters_meta_info.difference(
+                                 keys_parameters)))
+
+        self.external_version = external_version
+        self.uploader = uploader
+
+        self.custom_name = custom_name
+        self.tags = tags if tags is not None else []
+        self.binary_url = binary_url
+        self.binary_format = binary_format
+        self.binary_md5 = binary_md5
+        self.version = version
+        self.upload_date = upload_date
+        self.language = language
+        self.dependencies = dependencies
+        self.flow_id = flow_id
+
+    def _to_xml(self):
         """Generate xml representation of self for upload to server.
 
         Returns
         -------
-        flow_xml : string
+        str
             Flow represented as XML string.
         """
-        model = self.model
+        flow_dict = self._to_dict()
+        flow_xml = xmltodict.unparse(flow_dict, pretty=True)
+
+        # A flow may not be uploaded with the xml encoding specification:
+        # <?xml version="1.0" encoding="utf-8"?>
+        flow_xml = flow_xml.split('\n', 1)[-1]
+        return flow_xml
 
-        flow_dict = OrderedDict()
-        flow_dict['oml:flow'] = OrderedDict()
-        flow_dict['oml:flow']['@xmlns:oml'] = 'http://openml.org/openml'
-        flow_dict['oml:flow']['oml:name'] = self._get_name()
-        flow_dict['oml:flow']['oml:external_version'] = self.external_version
-        flow_dict['oml:flow']['oml:description'] = self.description
+    def _to_dict(self):
+        """ Helper function used by _to_xml and _to_dict.
```

Comment:
No, this is correct. See the rest of the docstring:
```
Since a flow can contain subflows (components) this helper 
function calls itself recursively to also  serialize these flows
to dictionaries.
```